cpp/src/arrow/compute/exec/asof_join_node_test.cc

```diff
@@ -74,237 +226,723 @@ void CheckRunOutput(const BatchesWithSchema& l_batches,
+    std::uniform_int_distribution<size_t> l_distribution(0, l_types.size() - 1);
+    std::uniform_int_distribution<size_t> r0_distribution(0, r0_types.size() - 1);
+    std::uniform_int_distribution<size_t> r1_distribution(0, r1_types.size() - 1);
+
+    for (int i = 0; i < 1000; i++) {
+      auto time_type = time_types[time_distribution(engine)];
+      ARROW_SCOPED_TRACE("Time type: ", *time_type);
+      auto key_type = key_types[key_distribution(engine)];
+      ARROW_SCOPED_TRACE("Key type: ", *key_type);
+      auto l_type = l_types[l_distribution(engine)];
+      ARROW_SCOPED_TRACE("Left type: ", *l_type);
+      auto r0_type = r0_types[r0_distribution(engine)];
+      ARROW_SCOPED_TRACE("Right-0 type: ", *r0_type);
+      auto r1_type = r1_types[r1_distribution(engine)];
+      ARROW_SCOPED_TRACE("Right-1 type: ", *r1_type);
+
+      RunTypes({time_type, key_type, l_type, r0_type, r1_type}, batches_runner);
+
+      auto end_time = std::chrono::system_clock::now();
+      std::chrono::duration<double> diff = end_time - start_time;
+      if (diff.count() > 2) {
+        std::cerr << "AsofJoin test reached time limit at iteration " << i << std::endl;
+        // this normally happens on slow CI systems, but is fine
+        break;
+      }
+    }
+  }
+  template <typename BatchesRunner>
+  void RunTypes(BasicTestTypes basic_test_types, BatchesRunner batches_runner) {
+    const BasicTestTypes& b = basic_test_types;
+    auto l_schema =
+        schema({field("time", b.time), field("key", b.key), field("l_v0", b.l_val)});
+    auto r0_schema =
+        schema({field("time", b.time), field("key", b.key), field("r0_v0", b.r0_val)});
+    auto r1_schema =
+        schema({field("time", b.time), field("key", b.key), field("r1_v0", b.r1_val)});
+
+    auto exp_schema = schema({
+        field("time", b.time),
+        field("key", b.key),
+        field("l_v0", b.l_val),
+        field("r0_v0", b.r0_val),
+        field("r1_v0", b.r1_val),
+    });
+
+    // Test three table join
+    BatchesWithSchema l_batches, r0_batches, r1_batches, exp_nokey_batches, exp_batches;
+    l_batches = MakeBatchesFromNumString(l_schema, l_data);
+    r0_batches = MakeBatchesFromNumString(r0_schema, r0_data);
+    r1_batches = MakeBatchesFromNumString(r1_schema, r1_data);
+    exp_nokey_batches = MakeBatchesFromNumString(exp_schema, exp_nokey_data);
+    exp_batches = MakeBatchesFromNumString(exp_schema, exp_data);
+    batches_runner(l_batches, r0_batches, r1_batches, exp_nokey_batches, exp_batches);
+  }
+
+  std::vector<util::string_view> l_data;
+  std::vector<util::string_view> r0_data;
+  std::vector<util::string_view> r1_data;
+  std::vector<util::string_view> exp_nokey_data;
+  std::vector<util::string_view> exp_data;
+  int64_t tolerance;
+};
+
 class AsofJoinTest : public testing::Test {};
 
-TEST(AsofJoinTest, TestBasic1) {
+#define ASOFJOIN_TEST_SET(name, num)                           \
```

Comment:
Do you think it would be possible to make this a parameterized test instead of a macro?

Reply:
I'll check this out after pushing a commit in response to your other comments.

Reply:
I fixed this.